site/website/blog/2020-05-20-designing-evergreen-cache-cdc.md

```diff
@@ -0,0 +1,361 @@
+data, depending on the underlying disk technology.
+
+## Hazelcast Jet
+
+[Hazelcast Jet](https://jet-start.sh/) is a distributed stream
+processing framework built on Hazelcast and combines a cache with
+fault-tolerant data processing.
+It has sources and sinks to integrate with several file, messaging and
+database systems (such as Amazon S3, Kafka, message brokers and
+relational databases).
+
+Jet also provides a Debezium module where it can process change events
+directly from the database, and write them to its distributed key-value
+store.
+This avoids having to write the intermediate messages to Kafka and then
+read again to be written to a separate cache.
+
+## Putting it all together
+
+It’s (finally!) time to assemble all the previous bits together.
+Here are the components and their responsibilities:
+
+1. A MySQL database instance is where the data is stored.
+It’s accessed in read-only mode by the cache, and in write-only mode by
+some external component
+2. A Jet instance reads events to MySQL through the Debezium connector,
+transforms them into cache-compatible key-value pairs, and updates the
+cache accordingly.
+Note that while Jet pipelines provide filtering capabilities, it’s also
+possible to filter items in the CDC connector to optimize load of the
+pipeline
+3. The app uses the cache, which is always up-to-date with the database,
+ give or take the time it takes for the above to execute
+
+![Final architecture with
+CDC](assets/2020-05-20-architecture-with-cdc.svg)
+
+Note that this architecture assumes one starts from a legacy state, with
+an existing app that uses caching, where later on a new component was
+set up that could update the database.
+
+If one starts from scratch, it’s possible to simplify the above diagram
+(and associated code) as Jet embeds its own Hazelcast instance.
+In that case, instead of Jet being a client of a third-party Hazelcast
+instance, Jet is the one to configure and start the instance.
+Obviously, it also can then get/put data.
+
+## Talk is cheap, show me the code
+
+Sources for this post are available [on
+GitHub](https://github.com/hazelcast/evergreen-cache).
+
+The repository is made of the following modules:
+
+- `app` is a Spring Boot application using Spring Data JDBC to access a
+MySQL database.
+It abstracts away Hazelcast by using a Spring Cache layer
+- `update` is a Spring Shell application.
+It allows to update the data inside the database, with the cache none
+the wiser
+- `pipeline` is the Jet pipeline that listens to CDC events and update
+the cache when data is updated
+
+The pipeline definition is quite straightforward:
+
+```java
+pipeline.readFrom(DebeziumSources.cdc(configuration))           // 1
+  .withoutTimestamps()
+  .map(r -> Values.convertToString(r.valueSchema(), r.value())) // 2
+  .apply(new JsonToPerson())                                    // 3
+  .writeTo(
+    SinkBuilder.sinkBuilder(                                    // 4
```

Comment:
this should be replaced my remoteMap sink, to simplify the logic and also to prevent the memory leak(since client will not be shutdown automatically with this approach after job finishes)